third_party/move/move-compiler-v2/transactional-tests/tests/tests.rs

```diff
@@ -4,38 +4,115 @@
 
-/// Tests which should be run with the `no-safety` experiment on.
-const NO_SAFETY_PATH: &str = "/no-safety/";
-
-/// Experiment env var
-const MOVE_COMPILER_EXP: &str = "MOVE_COMPILER_EXP";
-
 fn move_test_debug() -> bool {
     static MOVE_TEST_DEBUG: Lazy<bool> = Lazy::new(|| read_bool_env_var("MOVE_TEST_DEBUG"));
     *MOVE_TEST_DEBUG
 }
 
-fn run(path: &Path) -> Result<(), Box<dyn std::error::Error>> {
+#[derive(Clone)]
+struct TestConfig {
+    name: &'static str,
+    runner: fn(&Path) -> datatest_stable::Result<()>,
+    experiments: &'static [(&'static str, bool)],
+}
+
+const TEST_CONFIGS: &[TestConfig] = &[
+    TestConfig {
+        name: "optimize",
+        runner: |p| run(p, get_config_by_name("optimize")),
+        experiments: &[(Experiment::OPTIMIZE, true)],
+    },
+    TestConfig {
+        name: "no-optimize",
+        runner: |p| run(p, get_config_by_name("no-optimize")),
+        experiments: &[(Experiment::OPTIMIZE, false)],
+    },
+    TestConfig {
+        name: "optimize-no-simplify",
+        runner: |p| run(p, get_config_by_name("optimize-no-simplify")),
+        experiments: &[
+            (Experiment::OPTIMIZE, true),
+            (Experiment::AST_SIMPLIFY, false),
+        ],
+    },
+];
+
+/// Test files which must use separate baselines because their result
+/// is different.
+const SEPARATE_BASELINE: &[&str] = &[
+    // Runs into too-many-locals or stack overflow if not optimized
+    "inlining/deep_exp.move",
+    "constants/large_vectors.move",
+    // Printing bytecode is different depending on optimizations
+    "no-v1-comparison/print_bytecode.move",
+];
+
+fn get_config_by_name(name: &str) -> TestConfig {
+    TEST_CONFIGS
+        .iter()
+        .find(|c| c.name == name)
+        .cloned()
+        .unwrap_or_else(|| panic!("undeclared test confio `{}`", name))
+}
+
+fn run(path: &Path, config: TestConfig) -> datatest_stable::Result<()> {
     logging::setup_logging_for_testing();
     let p = path.to_str().unwrap_or_default();
-    if p.contains(NO_SAFETY_PATH) {
-        std::env::set_var(MOVE_COMPILER_EXP, "no-safety")
-    }
-    let test_config = if p.contains(SKIP_V1_COMPARISON_PATH) || move_test_debug() {
-        TestRunConfig::CompilerV2
+    let exp_suffix = if SEPARATE_BASELINE.iter().any(|s| p.contains(s)) {
+        Some(format!("{}.exp", config.name))
+    } else {
+        None
+    };
+    let v2_experiments = config
+        .experiments
+        .iter()
+        .map(|(s, v)| (s.to_string(), *v))
+        .collect_vec();
+    let vm_test_config = if p.contains(SKIP_V1_COMPARISON_PATH) || move_test_debug() {
+        TestRunConfig::CompilerV2 { v2_experiments }
     } else {
-        TestRunConfig::ComparisonV1V2
+        TestRunConfig::ComparisonV1V2 { v2_experiments }
     };
-    vm_test_harness::run_test_with_config(test_config, path)
+    vm_test_harness::run_test_with_config_and_exp_suffix(vm_test_config, path, &exp_suffix)
+}
+
+fn main() {
```

Comment:
Sorry for being stupid, but what does `main` do in this context?  Why didn't we need it before?  Feel free to just tell me the relevant crate name to go find docs.

Reply:
The `main` was generated before by the `datatest_stable::harness!(run_test, TEST_DIR, r".*\.(mvir|move)$")` macro (example of use). Now we are doing it manually and build up the list of tests ourselves. `main` is really the entrypoint for tesing here.